stl/inc/span

```diff
@@ -0,0 +1,727 @@
+concept _Is_span_compatible_range =
+    !is_array_v<remove_cvref_t<_Rng>>
+ && !_Is_span_v<remove_cvref_t<_Rng>>
+ && !_Is_std_array_v<remove_cvref_t<_Rng>>
+ && _RANGES contiguous_range<_Rng>
+ && _RANGES sized_range<_Rng>
+ && (_RANGES safe_range<_Rng> || is_const_v<_Ty>)
+ && is_convertible_v<remove_reference_t<_RANGES range_reference_t<_Rng>>(*)[], _Ty(*)[]>;
+// clang-format on
+#else // ^^^ __cpp_lib_concepts / !__cpp_lib_concepts vvv
+
+// STRUCT TEMPLATE _Is_span
+template <class>
+struct _Is_span : false_type {};
+
+template <class _Ty, size_t _Extent>
+struct _Is_span<span<_Ty, _Extent>> : true_type {};
+
+// STRUCT TEMPLATE _Is_std_array
+template <class>
+struct _Is_std_array : false_type {};
+
+template <class _Ty, size_t _Size>
+struct _Is_std_array<array<_Ty, _Size>> : true_type {};
+
+// STRUCT TEMPLATE _Is_span_convertible_range
+template <class _Rng, class _Ty>
+struct _Is_span_convertible_range
+    : bool_constant<is_convertible_v<remove_pointer_t<decltype(_STD data(_STD declval<_Rng&>()))> (*)[], _Ty (*)[]>> {};
+
+// STRUCT TEMPLATE _Has_container_interface
+template <class, class = void>
+struct _Has_container_interface : false_type {};
+
+template <class _Rng>
+struct _Has_container_interface<_Rng,
+    void_t<decltype(_STD data(_STD declval<_Rng&>())), decltype(_STD size(_STD declval<_Rng&>()))>> : true_type {};
+
+// STRUCT TEMPLATE _Is_span_compatible_range
+// clang-format off
+template <class _Rng, class _Ty>
+struct _Is_span_compatible_range : bool_constant<conjunction_v<
+    negation<is_array<_Rng>>,
+    negation<_Is_span<remove_const_t<_Rng>>>,
+    negation<_Is_std_array<remove_const_t<_Rng>>>,
+    _Has_container_interface<_Rng>,
+    _Is_span_convertible_range<_Rng, _Ty>>> {};
+// clang-format on
+
+#endif // !__cpp_lib_concepts
+
+// [views.span]
+// CLASS TEMPLATE span
+template <class _Ty, size_t _Extent = dynamic_extent>
+class span : public _Span_extent_type<_Extent> {
+public:
+    using element_type           = _Ty;
+    using value_type             = remove_cv_t<_Ty>;
+    using size_type              = size_t;
+    using difference_type        = ptrdiff_t;
+    using pointer                = _Ty*;
+    using const_pointer          = const _Ty*;
+    using reference              = _Ty&;
+    using const_reference        = const _Ty&;
+    using iterator               = _Span_iterator<_Ty>;
+    using const_iterator         = _Span_iterator<const _Ty>;
+    using reverse_iterator       = _STD reverse_iterator<iterator>;
+    using const_reverse_iterator = _STD reverse_iterator<const_iterator>;
+
+    static constexpr size_type extent = _Extent;
+
+    // [span.cons] Constructors, copy, and assignment
+#ifdef __cpp_lib_concepts
+
+    // clang-format off
+    constexpr span() noexcept requires (_Extent == 0 || _Extent == dynamic_extent) = default;
+
+    template <_Is_span_compatible_iterator<element_type> _It>
+    constexpr span(_It _First, size_type _Count) noexcept // strengthened

```

Comment:
This ctor should `_Verify_offset` the iterator if that is possible. i.e. this death test should fail:

```
vector<int> a;
a.push_back(1);
span<int, 2> b(a.first(), 2);
```

and this one:
```
vector<int> a;
a.push_back(1);
span<int> b(a.first(), 2);
```

Reply:
Ditto 412.

Reply:
I actually asked about that but @CaseyCarter disagreed

> This requires a valid range check
> Would it make sense to do _Adl_verify_range(_First, _First + _Count);

https://github.com/microsoft/STL/pull/142/files#r356779311

Reply:
And I missed to form the `_First + _Count`

Reply:
I pushed the missing `_First + _Count` and it popped some more questions into my mind:

1. Should we `const` qualify the arguments? `constexpr span(const _It _First, const size_type _Count)`
2. Should the `_Adl_verify_range(_First, _Last);` of the other constructor be moved into the `#if _CONTAINER_DEBUG_LEVEL > 0` to be more symmetric to this constructor (even if it has no functional effect as  `_Adl_verify_range` has its own `#if _CONTAINER_DEBUG_LEVEL > 0` check inside)

Reply:
I think attempting to call `_Verify_offset` would be better than attempting to form the destination iterator, as it results in better diagnostics when the check fails. It would be something like:

```
if constexpr(_Offset_verifiable_v<_It>) {
    // also needs an integer overflow check, see _Get_unwrapped_n in <xutility>
    _First._Verify_offset(static_cast<_Iter_diff_t<_It>>(_Count));
}
```

Reply:
Hmmm or Casey's `_Get_unwrapped_n` solution in a following comment which already does this.

Reply:
`_Get_unwrapped_n `it is